contrib/server/kafka/server.go

```diff
@@ -0,0 +1,113 @@
+package kafka
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-kratos/kratos/v2/log"
+	"github.com/go-kratos/kratos/v2/transport"
+
+	"golang.org/x/sync/errgroup"
+)
+
+var _ transport.Server = (*Server)(nil)
+
+// Message is a message abstraction
+type Message interface {
+	Key() string
+	Value() []byte
+}
+
+// Handler is used to handle the received message
+type Handler interface {
+	Topic() string
+	Handle(message Message) error
+}
+
+// Consumer receives and uses the handler to process messages
+type Consumer interface {
+	Topic() string
+	RegisterHandler(handler Handler)
+	HasHandler(topic string) bool
+	Consume(ctx context.Context) error
+	Close() error
+}
+
+// Server is a Kafka server wrapper
+type Server struct {
+	consumers []Consumer
+	handlers  map[string]Handler
+	logger    log.Helper
+}
+
+// ServerOption is a Kafka server option.
+type ServerOption func(server *Server)
+
+// Consumers registers a set of consumers to the Server.
+func Consumers(consumers []Consumer) ServerOption {
+	return func(server *Server) {
+		server.consumers = consumers
+	}
+}
+
+// Handlers registers a set of handlers to the Server.
+func Handlers(handlers []Handler) ServerOption {
+	return func(server *Server) {
+		for _, handler := range handlers {
+			server.handlers[handler.Topic()] = handler
+		}
+	}
+}
+
+// NewServer creates a Kafka server by options.
+func NewServer(opts ...ServerOption) (transport.Server, error) {
+	server := &Server{handlers: make(map[string]Handler)}
+
+	for _, o := range opts {
+		o(server)
+	}
+
+	if len(server.consumers) == 0 {
+		return nil, fmt.Errorf("no consumers")
+	}
+	if len(server.handlers) == 0 {
+		return nil, fmt.Errorf("no handlers")
+	}
+
+	for _, consumer := range server.consumers {
+		for _, handler := range server.handlers {
+			if consumer.Topic() != handler.Topic() {
+				continue
+			}
+			consumer.RegisterHandler(handler)
```

Comment:
Why register handler at this point? I think someone can register the handler before passing the consumer.

Reply:
Yes, this way can't prevent the user register the handler themselves. The reason I write this way is I want to use wire to generate the NewServer code. I have a demo code here
```
// NewKafkaServer creates a new Kafka Server
func NewKafkaServer(config *conf.Server, handlers []kafka.Handler) (*kafka.Server, error) {
	// init consumers
	var consumers []kafka.Consumer
	brokers := config.Kafka.Brokers
	for _, cConf := range config.Kafka.Consumers {
		srvConsumer, err := consumer.NewGroupConsumer(brokers, cConf.Topics, cConf.Group)
		if err != nil {
			return nil, errors.WithMessage(err, "init consumer error")
		}
		consumers = append(consumers, srvConsumer)
	}

	return kafka.NewServer([]kafka.ServerOption{
		kafka.Consumers(consumers),
		kafka.Handlers(handlers),
	}...)
}

```

Reply:
Oh~I see.But I think wire does not allow multiple providers for one type as mentioned on this [doc](https://github.com/google/wire/blob/main/docs/faq.md#what-if-my-dependency-graph-has-two-dependencies-of-the-same-type)

Reply:
We could return a slice of handlers or consumers, I have a kafka-server-demo here https://github.com/EndlessIdea/kratos-kafka-server-demo.